RockPaperScissors/RockPaperScissors/main.swift

```diff
@@ -6,5 +6,148 @@
 
 import Foundation
 
-print("Hello, World!")
+var shouldRunGame = true
 
+var validInputRange: ClosedRange<Int> {
+    return 0...RockPaperScissors.allCases.count
+}
+
+enum Script: String, CustomStringConvertible {
+    var description: String {
+        self.rawValue
+    }
+    
+    case menu = "가위(1), 바위(2), 보(3)! <종료 : 0> : "
+    case win = "이겼습니다!"
+    case lose = "졌습니다!"
+    case draw = "비겼습니다!"
+    case gameEnd = "게임 종료"
+    case dismissedError = "잘못된 입력입니다. 다시 시도해주세요."
+    case choiceOutOfRangeError = "가위바위보와 매칭될 수 있는 유효값이 아닙니다."
+}
+
+enum RockPaperScissorsGameError: Error {
+    case invalidInput
+    case choiceOutOfRange
+}
+
+enum RockPaperScissorsJudgement {
+    case win
+    case lose
+    case draw
+}
+
+enum RockPaperScissors: CaseIterable {
+    // 가위 바위 보 외의 case가 추가된다면 게임 결과를 계산해줄 matchJudgement와 matchHandPose 또한 업데이트 해야합니다.
+    case rock
+    case paper
+    case scissors
+    
+    static func getRange() -> ClosedRange<Int> {
+        return 1...Self.allCases.count
+    }
+}
+
+func manageRockPaperScissorsGame() {
+    while shouldRunGame {
+        do {
+            try startRockPaperScissorsGame()
+        } catch RockPaperScissorsGameError.invalidInput {
+            print(Script.dismissedError)
+        } catch RockPaperScissorsGameError.choiceOutOfRange {
+            print(Script.choiceOutOfRangeError)
+        } catch {
+            print(error)
+        }
+    }
+    
+    print(Script.gameEnd)
+}
+
+func startRockPaperScissorsGame() throws {
+    while shouldRunGame {
+        print(Script.menu, terminator: "")
+        
+        let userInput = try receiveUserNumber()
+        if userInput == 0 {
+            shouldRunGame = false
+            break
+        }
+        
+        let choices = try initUserAndComputerChoice(by: userInput)
+        
+        let gameResult = gameJudgement(by: choices)
+        
+        switch gameResult {
+        case .win:
+            print(Script.win)
+            shouldRunGame = false
+        case .lose:
+            print(Script.lose)
+            shouldRunGame = false
+        case .draw:
+            print(Script.draw)
+        }
+    }
+}
+
+
+/// 사용자 입력을 받아서 타입을 변환하고 유효성을 검증하여 사용자 숫자를 반환합니다.
+/// - Throws: 잘못된 입력 오류
+/// - Returns: 사용자 숫자
```

Comment:
함수의 인터페이스를 읽어보면 `throws` 키워드가 적용되어 있으니 에러를 던질 수 있는 함수라는 점을 유추할 수 있을 것 같아요. 그래서 이 함수를 사용하는 사람 입장에서는 `throws`가 잘못된 입력 오류라는 사실보다 이 함수의 매개변수는 어떤 인자를 받을 수 있고, 어떤 상황에서 에러가 발생하는지에 대한 정보를 더 얻고 싶어 할 것 같아요.

```suggestion
/// 사용자 입력을 받아서 타입을 변환하고 유효성을 검증하여 사용자 숫자를 반환합니다.
/// 아무것도 입력하지 않은 상태로 입력을 확정, 정수로 변환할 수 없는 값을 입력, 유효한 범위 (x...y)의 정수값을 입력하지 않는 경우 에러가 발생합니다. 
/// - Returns: 사용자가 입력한 수를 정수 형태로 반환합니다.
```